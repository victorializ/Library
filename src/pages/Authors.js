import React, { useState } from 'react';

import { LoadingComponent, SelectedAuthor, AuthorsList } from '../components';
import { getAllAuthors } from '../services/http-client';
import { useRequest } from '../hooks';

function Authors() {
  const [ authorId, setAuthorId ] = useState(null);
  const response = useRequest(getAllAuthors);

  return (
    <div className='authors'>
      <div className='authors__list'>
        <LoadingComponent {...response}>
          <AuthorsList onAuthorSelect={authorId => setAuthorId(authorId)} data={response.data} />
        </LoadingComponent>
      </div>
      {authorId && <SelectedAuthor authorId={authorId} />}
    </div>
  );
}

export { Authors };