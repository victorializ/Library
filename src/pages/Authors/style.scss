@import '../../assets/variables.scss';

.authors {
    display: flex;
    align-items: top;
    &__list {
        overflow-y: scroll;
        height: 90vh;
        width: 300px;
        border: 1px solid #ccc;
        background-image: url(./../../assets/images/books.jpg);
        &-item {
            padding: $space-sm;
        }
    }
}