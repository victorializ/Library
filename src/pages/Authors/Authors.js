import React, { useState, useEffect } from 'react';

import { List, Segment } from 'semantic-ui-react';

import { BooksList, ErrorMessage } from '../../components';
import { constants } from '../../i18n';
import { getAllAuthors, getAuthor } from '../../services/http-client/authors';
import { useRequest, WithRequest } from '../../components';

import './style.scss';

const SelectedAuthorName = ({data: {firstName = '', lastName = ''}}) =>
  (firstName || lastName) && 
  <Segment className='authors__name'>{`${firstName} ${lastName}`}</Segment>

const AuthorsList = (
  {
    data,
    onAuthorSelect
  }) => 
  <List inverted divided>
    { data.map(
      ({
        authorId, 
        firstName, 
        lastName
      }) => 
      <List.Item key={authorId} onClick={() => onAuthorSelect(authorId)}>
        <List.Content 
          className='authors__list-item'>
          <List.Header>{`${firstName} ${lastName}`}</List.Header>
        </List.Content>
      </List.Item>
      )
    }
  </List>

const AuthorsBooks = ({data : {bookAuthors: books}}) =>
  books.length ? 
    <BooksList data={books.map(({book}) => book)} />
    : 
    <ErrorMessage text={constants.nobooks} />


const SelectedAuthor = ({authorId}) => {
  const selectedAuthor = useRequest(getAuthor, authorId);
  return (
    <div className='authors--selected'>
      <WithRequest 
        data={selectedAuthor[0]} 
        error={selectedAuthor[1]}
        WrappedComponent={SelectedAuthorName} 
      />
      <WithRequest 
        data={selectedAuthor[0]} 
        error={selectedAuthor[1]}
        WrappedComponent={AuthorsBooks} 
      />
    </div>
  ); 
}

function Authors() {
  const [ authorId, setAuthorId ] = useState(null);

  const authors = useRequest(getAllAuthors);

  return (
    <div className='authors'>
      <div className='authors__list'>
        <WithRequest 
          error={authors[1]}
          data={authors[0]} 
          onAuthorSelect={authorId => setAuthorId(authorId)}
          WrappedComponent={AuthorsList} 
        />
      </div>
      {authorId && <SelectedAuthor authorId={authorId} />}
    </div>
  );
}

export { Authors };