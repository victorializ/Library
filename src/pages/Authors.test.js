import React from 'react';
import { shallow } from 'enzyme';

import Authors from './Authors';

describe('Authors page component', () => {
    it('renders correctly', () => {
        expect(shallow(<Authors />)).toMatchSnapshot();
    }); 
});