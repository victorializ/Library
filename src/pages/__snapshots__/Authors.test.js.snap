// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Authors page component renders correctly 1`] = `
<div
  className="authors"
>
  <div
    className="authors__list"
  >
    <LoadingComponent
      data={null}
      error={
        Object {
          "message": null,
        }
      }
      loading={true}
    >
      <AuthorsList
        data={null}
        onAuthorSelect={[Function]}
      />
    </LoadingComponent>
  </div>
</div>
`;
